pallets/pallet-deposit-storage/src/deposit.rs

```diff
@@ -0,0 +1,169 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2023 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+use frame_support::{
+	sp_runtime::DispatchError,
+	traits::{
+		fungible::{hold::Mutate, Inspect},
+		tokens::Precision,
+	},
+};
+use kilt_support::Deposit;
+use pallet_dip_provider::{traits::ProviderHooks as DipProviderHooks, IdentityCommitmentVersion};
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::traits::Get;
+use sp_std::marker::PhantomData;
+
+use crate::{BalanceOf, Config, Error, HoldReason, Pallet};
+
+#[derive(Clone, Debug, Encode, Decode, Eq, PartialEq, Ord, PartialOrd, TypeInfo, MaxEncodedLen)]
+pub struct DepositEntry<AccountId, Balance, Reason> {
+	pub(crate) deposit: Deposit<AccountId, Balance>,
+	pub(crate) reason: Reason,
+}
+
+pub struct FixedDepositCollectorViaDepositsPallet<DepositsNamespace, FixedDepositAmount>(
+	PhantomData<(DepositsNamespace, FixedDepositAmount)>,
+);
+
+pub enum FixedDepositCollectorViaDepositsPalletError {
+	DepositAlreadyTaken,
+	DepositNotFound,
+	FailedToHold,
+	FailedToRelease,
+	Internal,
+}
+
+impl From<FixedDepositCollectorViaDepositsPalletError> for u16 {
+	fn from(value: FixedDepositCollectorViaDepositsPalletError) -> Self {
+		match value {
+			FixedDepositCollectorViaDepositsPalletError::DepositAlreadyTaken => 0,
+			FixedDepositCollectorViaDepositsPalletError::DepositNotFound => 1,
+			FixedDepositCollectorViaDepositsPalletError::FailedToHold => 2,
+			FixedDepositCollectorViaDepositsPalletError::FailedToRelease => 3,
+			FixedDepositCollectorViaDepositsPalletError::Internal => u16::MAX,
+		}
+	}
+}
+
+impl<Runtime, DepositsNamespace, FixedDepositAmount> DipProviderHooks<Runtime>
```

Comment:
Somehow it feels weird to see this tight coupling of the deposits and the dip pallet here. Shouldn't this logic also be passed in while configuring the pallet for a specific runtime like with the removal?

Reply:
Do you think the name of the type is misleading? This is already a concrete implementation of a trait of the DIP provider pallet, and it is supposed to work only when you are storing identity commitments and want to take deposits for them. I don't understand what you refer with the logic being passed? Can you elaborate on that?

Reply:
I think the question is more from an architectural point of view: Why does something that is obviously specific to the DIP pallet (like this implementation of the DipProviderHooks) living within the deposit-storage pallet? Consequently this would mean that the deposit pallet code base would need to grow with every new "consumer" pallet that uses the deposit pallet. I think no action item should result from this comment, I'm only interested in your reasoning for structuring it this way.

Reply:
The problem is the Rust orphan rule. You can only implement a trait for a type if either of them is native to the crate. In this case, it would mean either implementing the trait in the pallet storage or in the pallet provider crate, as I see no point in having it in any other place. Would you rather have the type in the provider crate?